ee/draw/samples/vu1/vu1.c

```diff
@@ -0,0 +1,164 @@
+/*
+# _____     ___ ____     ___ ____
+#  ____|   |    ____|   |        | |____|
+# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
+#-----------------------------------------------------------------------
+# (c) 2020 Sandro Sobczyński <sandro.sobczynski@gmail.com>
+# Licenced under Academic Free License version 2.0
+# Review ps2sdk README & LICENSE files for further details.
+#
+*/
+
+#include <dma.h>
+#include <stdio.h>
+#include <malloc.h>
+#include <string.h>
+#include <kernel.h>
+#include <packet2.h>
+#include "vu1.h"
+
+static u32 vu1_inserted_qwords;
+
+// ----
+// Main
+// ----
+
+void vu1_send_data(u32 t_dest_address, void *t_data, u32 t_quad_size)
+{
+    packet2_t *packet2 = packet2_create_chain(2, P2_TYPE_NORMAL, 1);
+    packet2_chain_ref(packet2, t_data, t_quad_size, 0, 0, 0);
+    packet2_vif_stcycl(packet2, 0, 0x0101, 0);
+    packet2_vif_open_unpack(packet2, P2_UNPACK_V4_32, t_dest_address, 0, 0, 0, 0);
+    packet2_vif_close_unpack(packet2, t_quad_size);
+    packet2_chain_open_end(packet2, 0, 0);
+    packet2_vif_nop(packet2, 0);
+    packet2_vif_nop(packet2, 0);
+    packet2_chain_close_tag(packet2);
+    dma_channel_send_packet2(packet2, DMA_CHANNEL_VIF1, 1);
+    dma_channel_wait(DMA_CHANNEL_VIF1, 0);
+    packet2_free(packet2);
+}
+
+void vu1_set_double_buffer(u16 base, u16 offset)
+{
+    packet2_t *packet2 = packet2_create_chain(1, P2_TYPE_NORMAL, 1);
+    packet2_chain_open_end(packet2, 0, 0);
+    packet2_vif_base(packet2, base, 0);
+    packet2_vif_offset(packet2, offset, 0);
+    packet2_chain_close_tag(packet2);
+    dma_channel_wait(DMA_CHANNEL_VIF1, 0);
+    dma_channel_send_packet2(packet2, DMA_CHANNEL_VIF1, 1);
+    dma_channel_wait(DMA_CHANNEL_VIF1, 0);
+    packet2_free(packet2);
+}
+
+void vu1_add_data(packet2_t *packet2, void *t_data, u32 t_size, u8 t_use_top)
+{
+    packet2_chain_ref(packet2, t_data, t_size, 0, 0, 0);
+    packet2_vif_stcycl(packet2, 0, 0x0101, 0);
+    packet2_vif_open_unpack(packet2, P2_UNPACK_V4_32, vu1_inserted_qwords >> 4, t_use_top, 0, 1, 0);
+    packet2_vif_close_unpack(packet2, t_size);
+    vu1_inserted_qwords += t_size * 8;
+}
+
+void vu1_add_start_program(packet2_t *packet2)
+{
+    packet2_chain_open_cnt(packet2, 0, 0, 0);
+    packet2_vif_mscal(packet2, 0, 0);
+    packet2_vif_flush(packet2, 0);
+    packet2_chain_close_tag(packet2);
+}
+
+void vu1_send_packet(packet2_t *packet2)
+{
+    packet2_chain_open_end(packet2, 0, 0);
+    packet2_vif_nop(packet2, 0);
+    packet2_vif_nop(packet2, 0);
+    packet2_chain_close_tag(packet2);
+    dma_channel_send_packet2(packet2, DMA_CHANNEL_VIF1, 1);
+    dma_channel_wait(DMA_CHANNEL_VIF1, 0);
```

Comment:
We can't fix this in an easy way, but waiting = killing performance. A rewrite of libdma would be needed. We would need a queue where we can add things into without having to wait.
libdma could even use the "chain" feature to add new transfers into the chain, instead of into the queue.

This is actually what linux DMA drivers on other platforms do, and also what this PS2 TGE code seems to do:
https://github.com/rickgaiser/kernelloader/blob/ce0fb430a23660772f51de46e726dc34ccb915a6/TGE/sbios/sifdma.c#L285